Content.Server/Roles/RoleTimerSystem.cs

```diff
@@ -0,0 +1,306 @@
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly IPlayerManager _playerManager = default!;
+    [Dependency] private readonly GameTicker _ticker = default!;
+    [Dependency] private readonly RoleTimerManager _roleTimers = default!;
+
+    /// <summary>
+    /// Autosave regularly in case of server crash.
+    /// </summary>
+    private float _autosaveDelay = 900;
+
+    private float _autoSaveAccumulator;
+
+    /// <summary>
+    /// If someone just joined track the last time we set their times so the autosave doesn't round up.
+    /// </summary>
+    private readonly Dictionary<IPlayerSession, TimeSpan> _lastSetTime = new();
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        ConfigManager.OnValueChanged(CCVars.GameRoleTimersSaveFrequency, SetAutosaveDelay, true);
+        _playerManager.PlayerStatusChanged += OnPlayerStatusChanged;
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(OnRoundEnd);
+        SubscribeLocalEvent<PlayerAttachedEvent>(OnPlayerAttached);
+        SubscribeLocalEvent<PlayerDetachedEvent>(OnPlayerDetached);
+        SubscribeLocalEvent<RoleAddedEvent>(OnRoleAdd);
+        SubscribeLocalEvent<RoleRemovedEvent>(OnRoleRemove);
+        SubscribeLocalEvent<AFKEvent>(OnAFK);
+        SubscribeLocalEvent<UnAFKEvent>(OnUnAFK);
+    }
+
+    public IEnumerable<string> GetTimedRoles(Mind.Mind mind)
+    {
+        foreach (var role in mind.AllRoles)
+        {
+            if (role is not IRoleTimer timer) continue;
+            yield return ProtoManager.Index<RoleTimerPrototype>(timer.Timer).ID;
+        }
+    }
+
+    private List<string> GetTimedRoles(IPlayerSession session)
+    {
+        var roles = new List<string>();
+        var contentData = _playerManager.GetPlayerData(session.UserId).ContentData();
+
+        if (contentData?.Mind == null) return roles;
+
+        foreach (var mindRole in GetTimedRoles(contentData.Mind))
+        {
+            roles.Add(mindRole);
+        }
+
+        return roles;
+    }
+
+    private void OnRoleRemove(RoleRemovedEvent ev)
+    {
+        if (ev.Mind.Session == null) return;
+        Save(ev.Mind.Session, _timing.CurTime);
+    }
+
+    private void OnRoleAdd(RoleAddedEvent ev)
+    {
+        if (ev.Mind.Session == null) return;
+
+        var time = "";
+        if (ev.Role is IRoleTimer timer)
+            time = ProtoManager.Index<RoleTimerPrototype>(timer.Timer).ID;
+
+        // Save all but the current role.
+        SaveRoles(ev.Mind.Session, _timing.CurTime, GetTimedRoles(ev.Mind).Where(r => r != time));
+    }
+
+    private void SetAutosaveDelay(float value) => _autosaveDelay = value;
+
+    public override void Shutdown()
+    {
+        base.Shutdown();
+        ConfigManager.UnsubValueChanged(CCVars.GameRoleTimersSaveFrequency, SetAutosaveDelay);
+        _playerManager.PlayerStatusChanged -= OnPlayerStatusChanged;
+        FullSave();
```

Comment:
Entity systems should not be in charge of making sure that things are correctly saved on shutdown. Such data management operations should be relegated to managers.